crates/expressions/tests/validation_test.rs

```diff
@@ -0,0 +1,204 @@
+use std::collections::HashMap;
+
+use spin_expressions::{provider::ProviderVariableKind, Key, Provider, ProviderResolver};
+use spin_locked_app::Variable;
+
+#[derive(Default)]
+struct ResolverTester {
+    providers: Vec<Box<dyn Provider>>,
+    variables: HashMap<String, Variable>,
+}
+
+impl ResolverTester {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    fn with_dynamic_provider(mut self) -> Self {
```

Comment:
The change to bake provider kinds into the resolver seems unlovely - I appreciate the goal of making the tests easier to read but for me this begins to move into then territory of "hiding what we are actually testing."  (It also seems to rule out a category of tests along the lines of "a single provider has one variable but not another".)